querqy-core/src/test/java/querqy/rewrite/QuerqyTemplateEngineTest.java

```diff
@@ -0,0 +1,130 @@
+    }
+
+    @Test(expected = TemplateParseException.class)
+    public void testErrorForNonMatchingParams() throws TemplateParseException, IOException {
+        new QuerqyTemplateEngine(resource("templating/error/non-matching-params.txt"));
+    }
+
+    @Test(expected = TemplateParseException.class)
+    public void testErrorForNonMatchingParams2() throws TemplateParseException, IOException {
+        new QuerqyTemplateEngine(resource("templating/error/non-matching-params-2.txt"));
+    }
+
+    @Test
+    public void testParseParameters() throws TemplateParseException, IOException {
+        QuerqyTemplateEngine querqyTemplateEngine = new QuerqyTemplateEngine(new StringReader(""));
+
+        Map<String, String> parameters;
+
+        parameters = querqyTemplateEngine.parseParameters(" a = 1 ");
+        assertThat(parameters).containsExactly(
+                new AbstractMap.SimpleEntry<>("a", "1"));
+
+        parameters = querqyTemplateEngine.parseParameters("a = 1 || b=2");
+        assertThat(parameters).containsExactly(
+                new AbstractMap.SimpleEntry<>("a", "1"),
+                new AbstractMap.SimpleEntry<>("b", "2"));
+    }
+
+    @Test
+    public void testRendering() throws IOException, TemplateParseException {
+        QuerqyTemplateEngine querqyTemplateEngine = new QuerqyTemplateEngine(
+                resource("templating/input-rendering.txt"));
+
+        assertThat(list(querqyTemplateEngine.renderedRules.reader))
+                .isEqualTo(list(resource("templating/expected-rendering.txt")));
+    }
+
+    @Test
+    public void testLineNumberMapping() throws IOException, TemplateParseException {
+        List<String> linesOfInput = list(resource("templating/input-line-number-mapping.txt"));
+        Map<Integer, String> numberedLinesOfInput = numberedLines(linesOfInput);
+
+        QuerqyTemplateEngine querqyTemplateEngine = new QuerqyTemplateEngine(
+                resource("templating/input-line-number-mapping.txt"));
+
+        List<String> linesOfOutput = list(querqyTemplateEngine.renderedRules.reader);
+        Map<Integer, String> numberedLinesOfOutput = numberedLines(linesOfOutput);
+
+        Map<Integer, Integer> lineNumberMapping = querqyTemplateEngine.renderedRules.lineNumberMapping;
+
+        assertThat(numberedLinesOfOutput.get(3)).isEqualTo(numberedLinesOfInput.get(lineNumberMapping.get(3)));
+        assertThat(numberedLinesOfOutput.get(7)).isEqualTo(numberedLinesOfInput.get(lineNumberMapping.get(7)));
+        assertThat(numberedLinesOfOutput.get(12)).isEqualTo(numberedLinesOfInput.get(lineNumberMapping.get(12)));
+        assertThat(numberedLinesOfOutput.get(19)).isEqualTo(numberedLinesOfInput.get(lineNumberMapping.get(19)));
+    }
+
+    private Reader resource(String resourceName) {
+        return new InputStreamReader(getClass().getClassLoader().getResourceAsStream(resourceName));
+    }
+
+    private void printReader(Reader reader) throws IOException {
```

Comment:
It seems that this method is never called (and we don't want System.out.println())